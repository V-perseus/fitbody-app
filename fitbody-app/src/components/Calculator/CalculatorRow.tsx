import React from 'react'
import { TextInput, Text, StyleSheet, View } from 'react-native'
// Assets
import globals from '../../config/globals'
import { HeightUnit, WeightUnit } from '../../data/user/types'
import CalculatorButton from './CalculatorButton'

export type RowValueType = { value: number | undefined; unit: HeightUnit | WeightUnit | undefined }
export type RowButtonsDataType = { value: HeightUnit | WeightUnit; title: string }
interface ICalculatorRowProps {
  type: 'height' | 'weight' | 'age' | 'activity'
  title: string
  value: RowValueType
  update: (value: RowValueType) => void
  activityValue?: string
  activityUpdate?: () => void
  inputData: string[]
  buttonsData: RowButtonsDataType[]
  updateUnit?: (unit: HeightUnit | WeightUnit) => void
  singleButton?: boolean
}
function CalculatorRow({
  type,
  title,
  value,
  activityValue,
  update,
  activityUpdate,
  buttonsData,
  updateUnit,
}: ICalculatorRowProps): JSX.Element {
  // Inches -> Cm
  // const convertInchesToCm = (inches) => parseFloat((inches * 2.54).toFixed(2))

  // // Cm -> Inches
  // const convertCmToInches = (cm) => parseFloat((cm / 2.54).toFixed(2))

  // // Pound -> Kg
  // const convertPoundToKg = (pound) => parseFloat((pound / 2.2046).toFixed(2))

  // // Kg -> Pound
  // const convertKgToPound = (kg) => parseFloat((kg * 2.2046).toFixed(2))

  function handleOnChangeText(text: string) {
    update({
      value:
        value.unit === 'inches' && value?.value !== undefined ? parseInt(text || '0') * 12 + (value.value % 12) : parseInt(text || '0'),
      unit: value.unit,
    })
  }

  function handleOnChangeTextInches(text: string) {
    update({
      value: Math.floor(value.value! / 12) * 12 + (parseFloat(text) || 0 / 12),
      unit: value.unit,
    })
  }

  function handleActivityUpdate() {
    activityUpdate?.()
  }

  let style = type === 'height' && value.unit === 'inches' ? styles.doubleInput : styles.singleInput

  let valueString = value ? value.value?.toString() : '0'

  return (
    <View style={styles.row}>
      <Text style={styles.label}>{title}</Text>
      <View style={styles.internalRow}>
        {/* Field */}
        {type === 'activity' ? (
          <>
            <TextInput
              // placeholder={placeholder}
              testID="input"
              style={[styles.input, styles.singleInput]}
              value={activityValue}
              autoCorrect={false}
            />
            <CalculatorButton key={'single_btn'} title={'UPDATE'} handlePress={handleActivityUpdate} active={true} />
          </>
        ) : (
          <>
            <TextInput
              // placeholder={placeholder}
              testID="input"
              style={[styles.input, style]}
              value={value.unit === 'inches' && value?.value !== undefined ? Math.floor(value.value / 12).toString() : valueString}
              onChangeText={handleOnChangeText}
              autoCorrect={false}
              keyboardType="number-pad"
              returnKeyType="done"
            />
            {/* Inches field */}
            {value.unit === 'inches' && value.value !== undefined ? (
              <TextInput
                // placeholder={inputData[1]}
                testID="inches_input"
                style={[styles.input, styles.doubleInput]}
                value={value.value ? Math.floor(value.value % 12).toString() : undefined}
                onChangeText={handleOnChangeTextInches}
                autoCorrect={false}
                keyboardType="number-pad"
                returnKeyType="done"
              />
            ) : null}

            {/* Buttons */}
            {buttonsData.map((item, index) => (
              <CalculatorButton
                key={`cb${index}`}
                title={item.title}
                handlePress={() => updateUnit?.(item.value)}
                active={item.value === value.unit}
              />
            ))}
          </>
        )}
      </View>
    </View>
  )
}

export default CalculatorRow

const styles = StyleSheet.create({
  row: {
    flex: 1,
    flexDirection: 'column',
    width: globals.window.width,
    paddingLeft: 15,
    paddingRight: 15,
    marginTop: 20,
    marginBottom: 20,
  },
  label: {
    color: globals.styles.colors.colorBlack,
    fontSize: 16,
    fontFamily: globals.fonts.primary.semiboldStyle.fontFamily,
  },
  internalRow: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 20,
  },
  input: {
    fontSize: 14,
    color: globals.styles.colors.colorBlack,
    borderBottomColor: globals.styles.colors.colorGray,
    borderBottomWidth: 1,
    paddingTop: 2,
    paddingBottom: 13,
  },
  singleInput: {
    width: globals.window.width * 0.44,
  },
  doubleInput: {
    width: globals.window.width * 0.22,
  },
})
