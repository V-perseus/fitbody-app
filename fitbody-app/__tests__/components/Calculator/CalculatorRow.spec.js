import React from 'react'
import { fireEvent, render } from '@testing-library/react-native'

import CalculatorRow from '../../../src/components/Calculator/CalculatorRow'

describe('CalculatorRow', () => {
  const updateMock = jest.fn()
  const updateUnitMock = jest.fn()

  let component
  beforeEach(() => {
    component = render(
      <CalculatorRow
        type="height"
        value={{
          value: 5,
          unit: 'feet',
        }}
        title={'How tall are you?'}
        buttonsData={[
          { value: 'inches', title: 'feet' },
          { value: 'cm', title: 'centimeter' },
        ]}
        update={updateMock}
        updateUnit={updateUnitMock}
      />,
    )
  })

  it('renders snapshot correctly', () => {
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render row', () => {
    const { queryByText } = component
    expect(queryByText('How tall are you?')).not.toBeNull()
  })

  it('should handle input', () => {
    const { queryByTestId } = component
    fireEvent.changeText(queryByTestId('input'), 20)
    expect(updateMock).toBeCalledWith({ unit: 'feet', value: 20 })
    // expect(queryByText('CENTIMETER')).not.toBeNull()
  })

  it('should render and handle inches input', async () => {
    const { queryByTestId } = render(
      <CalculatorRow
        type="height"
        value={{
          value: 5,
          unit: 'inches',
        }}
        title={'How tall are you?'}
        buttonsData={[
          { value: 'inches', title: 'feet' },
          { value: 'cm', title: 'centimeter' },
        ]}
        update={updateMock}
        updateUnit={updateUnitMock}
      />,
    )
    fireEvent.changeText(queryByTestId('inches_input'), 20)
    expect(updateMock).toBeCalledWith({ unit: 'inches', value: 20 })
  })
})
