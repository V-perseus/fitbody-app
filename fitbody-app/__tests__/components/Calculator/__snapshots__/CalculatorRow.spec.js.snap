// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatorRow renders snapshot correctly 1`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "column",
      "marginBottom": 20,
      "marginTop": 20,
      "paddingLeft": 15,
      "paddingRight": 15,
      "width": 750,
    }
  }
>
  <Text
    style={
      {
        "color": "#333333",
        "fontFamily": "OpenSans-Semibold",
        "fontSize": 16,
      }
    }
  >
    How tall are you?
  </Text>
  <View
    style={
      {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 20,
        "marginTop": 20,
      }
    }
  >
    <TextInput
      autoCorrect={false}
      keyboardType="number-pad"
      onChangeText={[Function]}
      returnKeyType="done"
      style={
        [
          {
            "borderBottomColor": "#dedede",
            "borderBottomWidth": 1,
            "color": "#333333",
            "fontSize": 12,
            "paddingBottom": 15,
            "paddingTop": 2,
          },
          {
            "width": 330,
          },
        ]
      }
      testID="input"
      value="5"
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "#bbbbbb",
          "borderRadius": 10,
          "borderWidth": 1,
          "flexDirection": "column",
          "height": 20,
          "justifyContent": "center",
          "opacity": 1,
          "width": 150,
        }
      }
      testID="calc_btn"
    >
      <Text
        style={
          {
            "color": "#bbbbbb",
            "fontSize": 10,
          }
        }
      >
        feet
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "#bbbbbb",
          "borderRadius": 10,
          "borderWidth": 1,
          "flexDirection": "column",
          "height": 20,
          "justifyContent": "center",
          "opacity": 1,
          "width": 150,
        }
      }
      testID="calc_btn"
    >
      <Text
        style={
          {
            "color": "#bbbbbb",
            "fontSize": 10,
          }
        }
      >
        centimeter
      </Text>
    </View>
  </View>
</View>
`;
